### ISSUE TITLE
Tensorflow 2.0 too slow when minimizing a custom cost function

### ISSUE THREAD

**Issue by HTCode** [tensorflow/tensorflow#34144]
**System information**
- OS : Windows 10
- TensorFlow version: v2.0.0-rc2-26-g64c3d382ca 2.0.0
- Python version: 3.7.4

I have a code that looks like the following, where I want to minimize a custom cost function with respect to parameters w. However, when running the code, it appears to me that it is very slow (like more than 30 times slower) compared to the same code implemented without tensorflow (by explicitly defining a function that gives the gradient of the cost).

I am not sure if it's a problem with tf or if I am doing something wrong and unnecessarily re-computing the tf graph each time. I posted this issue on stackoverflow and have been advised to open an issue here.

In the following code, I am using a simple dummy cost function just as an example to show the big difference in performance.

**Code with Tensorflow:**

```
import numpy as np
import tensorflow as tf
import time

class ExampleTF:
    def __init__(self, n=100, m=10):
        Z = np.random.randn(n, m)
        self.Z = tf.convert_to_tensor(Z, dtype=tf.float32)
        self.w = tf.Variable(np.ones((m, 1)), dtype=tf.float32)

    # =====================================
    def cost(self, P):
        # This is a simple dummy cost function just as an example
        return tf.reduce_sum((self.Z @ self.w) - P)

    # =====================================
    def optimize_w(self, cost_func, parameters, lr=0.01, iterations=2000):
        optimizer = tf.optimizers.Adam(lr)
        for _ in range(iterations):
            optimizer.minimize(cost_func, var_list=parameters)

    # =====================================
    def update(self, P):
        P = tf.convert_to_tensor(P, dtype=tf.float32)

        self.optimize_w(
            cost_func = lambda: self.cost(P),
            parameters = [self.w]
        )

        #print("===> cost:", self.cost(P).numpy())
        #print("w:", self.w.numpy().reshape(-1)[:10])

# =====================================
n, m = 1000, 100
ex_tf = ExampleTF(n, m)
for _ in range(50):
    P = np.random.uniform(size=n).reshape((-1, 1))

    start = time.time()
    ex_tf.update(P)
    elapsed = time.time() - start

    print("elapsed time:", elapsed)
```

**Code without Tensorflow (just numpy) :**

```
import numpy as np
import tensorflow as tf
import time

class ExampleNonTF:
    def __init__(self, n=100, m=10):
        self.Z = np.random.randn(n, m)
        self.w = np.ones((m, 1))

    # =====================================
    def cost(self, P):
        # This is a simple dummy cost function just as an example
        return np.sum(self.Z @ self.w - P)

    # =====================================
    def gradient_cost(self, P):
        # This is the gradient of the dummy cost function with respect to self.w
        return np.sum(self.Z, axis=0).reshape(self.w.shape)

    # =====================================
    def optimize_w(self, P, lr=0.01, iterations=2000): # This is the ADAM optimizer
        avg_grad1 = 0; avg_grad2 = 0
        beta1 = 0.9; beta2 = 0.999; eps = 1e-07
        for itr in range(iterations):
            grad = self.gradient_cost(P)
            avg_grad1 = beta1 * avg_grad1 + (1 - beta1) * grad
            avg_grad2 = (beta2 * avg_grad2 + (1 - beta2) * (grad ** 2))
            avg_grad1_corr = avg_grad1 / (1 - beta1 ** (itr + 1))
            avg_grad2_corr = avg_grad2 / (1 - beta2 ** (itr + 1))
            self.w = self.w - lr * (avg_grad1_corr / (np.sqrt(avg_grad2_corr) + eps))

    # =====================================
    def update(self, P):
        self.optimize_w(P)

        #print("===> cost:", self.cost(P))
        #print("w:", self.w.reshape(-1)[:10])

# =====================================
n, m = 1000, 100
ex_nontf = ExampleNonTF(n, m)
for _ in range(50):
    P = np.random.uniform(size=n).reshape((-1, 1))

    start = time.time()
    ex_nontf.update(P)
    elapsed = time.time() - start

    print("elapsed time:", elapsed)
```

**Comment by ravikyram** [comment]
I have tried on colab with TF version 2.0  and was able to reproduce the issue.Please, find the gist [here](https://colab.sandbox.google.com/gist/ravikyram/4280a01f0727d3da964e6ad750043929/untitled346.ipynb). Thanks!

**Comment by HTCode** [comment]
By the way, for a very big n (e.g. > 1000000), the TF code version starts to become faster than the NON-TF one.

Is Tensorflow meant for large-scale computations only? I consider `n, m = 10000, 100` to be sufficiently big in my case, yet the NON-TF version is faster for this dimension.

**Comment by OverLordGoldDragon** [comment]
Started a [Colab](https://colab.research.google.com/drive/15jM5kLvd5YBn-dU5_u0cBNStHO9dMuGj) to organize performance tests. Noticed bit late that your Numpy implementation is off - see "Changelog". Lastly, note the very last optimistic result; as I suspected, `Eager` is a culprit - will continue investigating later. The notebook is editable - feel free to make changes.

**Comment by OverLordGoldDragon** [comment]
Ran some tests - code & results below. Noticed bit late that your Numpy implementation is off - see "Changelog". Lastly, note the very last optimistic result; as I suspected, `Eager` is a culprit - will continue investigating later.

<hr>

**Results**:

```python
EAGER
tf vs. np  (tf.optimizers.Adam)
_____________________________________________________
Ratio | n, m          | iterations | Total time (sec)
66.6  | 100,    100   | 50000      | (111,  1.67)
26.9  | 1000,   100   | 50000      | (120,  4.46)
4.36  | 10000,  100   | 20000      | (54.9, 12.6)
1.34  | 100000, 100   | 2000       | (25.6, 19.1)
1.02  | 100000, 1000  | 400        | (34.2, 33.6)
_____________________________________________________

tf.keras vs. np  (tf.keras.optimizers.Adam)
_____________________________________________________
Ratio | n, m          | iterations | Total time (sec)
64.9  | 100,    100   | 50000      | (114,  1.75)
25.4  | 1000,   100   | 50000      | (115,  4.53)
4.39  | 10000,  100   | 20000      | (55.3, 12.6)
1.26  | 100000, 100   | 2000       | (24.4, 19.4)
1.02  | 100000, 1000  | 400        | (33.9, 33.4)
_____________________________________________________

GRAPH
tf.keras vs. np  (tf.keras.optimizers.Adam)
_____________________________________________________
Ratio | n, m          | iterations | Total time (sec)
      | 100,    100   | 50000      | 
      | 1000,   100   | 50000      | 
      | 10000,  100   | 20000      | 
      | 100000, 100   | 2000       | 
0.373 | 100000, 1000  | 400        | (12.5, 33.6)
_____________________________________________________
```

<hr>

**Testing code**:

```python
import tensorflow as tf
# tf.compat.v1.disable_eager_execution() # uncomment to disable Eager
import numpy as np
from time import time
print("TF version: " + tf.__version__)

######################################################
class ExampleTF:
    def __init__(self, n=100, m=10, iterations=2000):
        Z = np.random.randn(n, m)
        self.Z = tf.convert_to_tensor(Z, dtype=tf.float32)
        self.w = tf.Variable(np.ones((m, 1)), dtype=tf.float32)
        self.iterations = iterations
        self.optimizer = tf.keras.optimizers.Adam(lr=0.01)

    def cost(self, P):
        return tf.reduce_sum((self.Z @ self.w) - P) # dummy cost function

    def optimize_w(self, cost_func, parameters, lr=0.01):
        for _ in range(self.iterations):
            self.optimizer.minimize(cost_func, var_list=parameters)

    def update(self, P):
        P = tf.convert_to_tensor(P, dtype=tf.float32)
        self.optimize_w(cost_func = lambda: self.cost(P),
                        parameters = [self.w])

class ExampleNumpy:
    def __init__(self, n=100, m=10, iterations=2000):
        self.Z = np.random.randn(n, m)
        self.w = np.ones((m, 1))
        self.iterations = iterations

    def cost(self, P):
        return np.sum(self.Z @ self.w - P)

    def gradient_cost(self, P):
        return np.sum(self.Z, axis=0).reshape(self.w.shape)

    def optimize_w(self, P, lr=0.01): # Adam implemented in Numpy
        m_t, v_t = 0, 0
        beta_1 = 0.9; beta_2 = 0.999; eps = 1e-07

        for itr in range(self.iterations):
            t = itr + 1
            lr_t = lr * np.sqrt(1 - beta_2**t) / (1 - beta_1**t)
            grad = self.gradient_cost(P)

            m_t = beta_1 * m_t + (1 - beta_1) * grad
            v_t = beta_2 * v_t + (1 - beta_2) * (grad**2)
            self.w = self.w - lr_t * m_t / (np.sqrt(v_t) + eps)

    def update(self, P):
        self.optimize_w(P)

def test_performance(ex_class, P, test_name):
    iters = ex_class.iterations
    t0 = time()
    ex_class.update(P)
    iter_time = round((time() - t0) / iters, 7)

    print("[{}] Time/iter: {} sec ({} iters avg) -- Total time: {} sec".format(
          test_name, iter_time, iters, iter_time * iters))
    return iter_time
######################################################

n, m = 100000, 1000
iterations = 400

P = np.random.uniform(size=n).reshape((-1, 1))
ex_tf = ExampleTF(n, m, iterations)
ex_np = ExampleNumpy(n, m, iterations)

tf_iter_time = test_performance(ex_tf, P, 'tf')
np_iter_time = test_performance(ex_np, P, 'np')

print("[tf/np] Time/iter:", round(tf_iter_time / np_iter_time, 3), "(ratio)")
```

<hr>

**Changelog**:
   - Defined self.optimizer to eliminate redundant redefinitions w/ overheads
   - Rewrote Numpy's self.optimize_w to be consistent w/ TF/keras implementation (partly*)
   - Removed `range(50)`; redundant overhead w/ `P` generations; `iterations` is sufficient
   - Defined "test_performance" func for convenience
   - Formatting, var renaming, etc

*  `self.cost` is never used, and `gradient` is computed without it - should be fixed. 
  Also the dummy classes should be such that they're buildable as neural nets - i.e. 
  involving layers w/ weights (e.g. `Dense`), a loss function `(e.g. `'mae', 'mse'`), 
  and labels (e.g. `P`).

**Todo**:
   - Profile Graph (disable Eager execution) performance
   - Profile Numpy implementation accuracy vs. TF

**Comment by cshjin** [comment]
@OverLordGoldDragon  
I don't know whether your optimization is completed on GPU or CPU. 
I find the same issue on GPU, but if you force the optimization in CPU, which is the same as numpy does, then it is faster.

On GPU:
```bash
TF version: 2.0.0
[tf] Time/iter: 0.0888078 sec (400 iters avg) -- Total time: 35.523120000000006 sec
[np] Time/iter: 0.0664769 sec (400 iters avg) -- Total time: 26.590760000000003 sec
[tf/np] Time/iter: 1.336 (ratio)
```

On CPU:
```bash
TF version: 2.0.0
[tf] Time/iter: 0.0362841 sec (400 iters avg) -- Total time: 14.51364 sec
[np] Time/iter: 0.0663887 sec (400 iters avg) -- Total time: 26.55548 sec
[tf/np] Time/iter: 0.547 (ratio)
```

I don't know the particular reasons behind, maybe the loading cost from CPU to GPU?

**Comment by jaingaurav** [comment]
There are known limitations of TF eager execution performance that is being actively worked on by the team. I'd suggest seeing if you can use tf.function to improve performance.
